### PR TITLE
feat(surveys): Add open-ended choices for multiple and single choice surveys

### PR DESCRIPTION
**Pull request by ssoonmi**
## Problem

Addresses this issue: https://github.com/PostHog/posthog/issues/17471

## Changes

- Added open-ended choices to multiple-choice and single-choice questions on
  surveys.
  - To keep the `choices` key on questions as an array of strings, I created
    logic to store and distinguish an open-ended choice from a regular choice
    with a string prefix of `"OPENlabel="`.
    - Ex: `choices = ['Yes', 'No', 'OPENlabel=Maybe']` would produce two regular
      choices (Yes and No) and one open-ended choice (Maybe)
  - Users can add more than one open-ended choice to a question.
  - Open-ended choices have flexible labeling, but the label defaults to
    "Other".
    - default: `"OPENlabel=Other"`
  - Design of the appearance of open-ended choices both was not clearly
    outlined, so I took creative liberty in the design and the user experience.
    Would love a review and feedback on the design and UX.
- Added "Add open-ended choice" button underneath the "Add choice" to the survey
  form when editing a multiple or single-choice survey question. Design of the
  button is identical to the existing "Add choice" button and is on the same
  row, but will flex to a new row for smaller screen sizes.
- Added Cypress tests for the form and appearance for open-ended choices on
  multiple-choice surveys.
- Changed frontend survey logic and logic tests for viewing single-choice survey
  results to zero-fill choices only if they aren't open-ended choices.
- Added frontend survey logic tests for viewing single-choice survey results
  with open-ended questions.
- Added validation for the backend on question choices data format.

Survey Appearance:
survey-appearance

Single-Choice Survey Form and Appearance:
single-choice

Multiple-Choice Survey Form and Appearance:
multiple-choice

Wrap buttons for small screen sizes:
wrap-buttons-small-screen-size

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

## How did you test this code?

- Cypress: I ran the changed `cypress/e2e/surveys.cy.ts` test file in Cypress
  and it passed in both Chrome and Microsoft Edge.
- Frontend: I ran the changed `frontend/src/scenes/surveys/surveyLogic.test.ts`
  test file using the command
  `pnpm jest --testPathPattern=frontend/src/scenes/surveys` and it passed.
- Backend: I ran the changed `posthog/api/test/test_survey.py` test file using
  the command `pytest posthog/api/test/test_survey.py` and it passed.
- Manual end-to-end testing: I wasn't sure how to actually take a survey after
  it's been created and launched. Because of this, I couldn't test the view for
  results of the survey. I would appreciate advice on how I could test this!
- Manual frontend testing: Here are some screen recordings of my manual frontend
  testing:

https://github.com/PostHog/posthog/assets/40774865/646ee6de-74a9-43f8-b792-3d8eac26466d

https://github.com/PostHog/posthog/assets/40774865/63959d4c-2c4d-40d5-a459-137da6d2e055

### CODE REVIEW

**Comment by ssoonmi** [comment]
Updated the survey appearance in posthog-js:
https://github.com/PostHog/posthog-js/pull/901

**Comment by neilkakkar** [comment]
hi @ssoonmi this is great! :)

One thing I'd mention is we don't need to keep the open ended choices as a custom string. We can instead add another parameter, called `open_ended_choices`, which is a list of labels. This should make the code cleaner, and get rid of all string slicing & unintended clashes.

On the product side, I'm not sure when multiple open ended choices make sense - will let @liyiy answer this. Right now, it _seems_ one is sufficient, in which case we can refine the data model to be even simpler.

**Comment by ssoonmi** [comment]
@neilkakkar Great! So if it is confirmed that y'all only need one open-ended choice per question, would the field be singular instead of plural (`open_ended_choice`)?

**Comment by ssoonmi** [comment]
There's one critical thing I haven't tested yet:

I couldn't test the UI for the summary of the survey results. I created a business logic test for the survey results with mocked survey result data, but I couldn't actually see the submitted survey results in the summary survey results view rendered in the `frontend/src/scenes/surveys/SurveyView.tsx` file.

**Comment by liyiy** [comment]
+1 on the restriction for only 1 open ended choice per question and having it always be at the bottom of the choices list

File: frontend/src/scenes/surveys/surveyLogic.test.ts
Comment (liyiy) [inline frontend/src/scenes/surveys/surveyLogic.test.ts:63]: I think we might start requiring more information to be added to our choices than an array of strings can handle. This is on us so I'm happy to update that part myself 😄 unless you really want to do it which is also awesome.

What I'm proposing is to add another property to the question object that's something like:
`choicesProperties: [{ choiceId: 2, openTextOptional: true }]` where it provides more details about the choice instead of having to rely on parsing the string for the `OPENlabel` prefix.

This could also be useful when we also implement logic branching survey questions that depend on specific choices, but basically it'd be nice to have some way of adding properties to answers.

cc @neilkakkar

Comment (neilkakkar) [reply frontend/src/scenes/surveys/surveyLogic.test.ts:63]: makes sense! Probably easier to always index by the id than having an array with key choiceId, so:

```
choiceProperties: {
 <id: number 0 to len(choices)-1>: { attributes }
 }
 ```
 
 .. although, it's a bit janky to keep these two different things in sync, that closely depend on each other, so it's worth deprecating this in the future and move to a choices array of objects, which has the label value, and all the other attributes.
 
 For example, if I delete a choice in the middle, the indexes of all choices after that choice need to change.
 
 So, I'd say go down this route when we do have the other attributes & deprecate the current format in one go; and in the meanwhile since the open choice is always going to be last + not common to all choice options, keep it as a separate thing that we can incorporate into the deprecation in the future.

Comment (liyiy) [reply frontend/src/scenes/surveys/surveyLogic.test.ts:63]: The additional attributes can always be added later on, so if we know we want to add `choiceProperties` and deprecate `choices` later on anyway, why not just add it now, skip the additional `open_ended_choices` field for the question, and only have the smallest addition of `isOpenTextOption`attribute?

I think this could be okay, since we'll only need to keep the "other" choice in sync with `choiceProperties` for now

Comment (liyiy) [reply frontend/src/scenes/surveys/surveyLogic.test.ts:63]: Unless you meant to refactor `choices` into an array of objects, which in that case, then we can hold off on `choiceProperties` and the string prefix format should be a fine placeholder for now

Comment (liyiy) [reply frontend/src/scenes/surveys/surveyLogic.test.ts:63]: The reason I also thought adding `choiceProperties` would be fine was that not all choices actually need additional attributes onto it 😁 So we would only need to store properties for certain choices, at the expense of just keeping track the ID index value when answer choices get deleted/added, seems like an okay tradeoff to me to reduce bloat on the survey itself

Comment (neilkakkar) [reply frontend/src/scenes/surveys/surveyLogic.test.ts:63]: Only reason to wait is to figure out what attributes we need & whether they will change how the data should be stored.

But can be ok to deprecate now as well if you think they'll fit reasonably well into an object configuration, your call!

I think this bit wasn't clear: If we do deprecate, I'd suggest not adding `choiceProperties`, but a new `choices` array that is:

```
extendedChoices: [
{ label: 'yes'},
{ label: 'no'},
{ label: 'maybe', isOpenEnded: true}
]
```

so there's nothing to keep in sync, and posthog-js adjusts to use these new choices if they exist, otherwise the old ones.

So choose between this^ & `open_ended_choice` attribute, let's not do `choiceProperties` as an extra attribute. Make sense?

Comment (neilkakkar) [reply frontend/src/scenes/surveys/surveyLogic.test.ts:63]: ah, saw your follow up comments later - yes I meant array of object choices.

> seems like an okay tradeoff to me to reduce bloat on the survey itself

I'm not sure it will make that much of a difference for bloat. Not too keen because it's a lot more bug-prone.

Comment (ssoonmi) [reply frontend/src/scenes/surveys/surveyLogic.test.ts:63]: Since there would only ever be one open-ended choice in a question and it would always be the last choice in the list of choices, another option could be to keep choices as a string and add an extra field on the question itself, such as `has_open_choice`, that would force the last choice in the choices to be an open-ended choice. That way, you won't need to worry as much about introducing lots of bugs with this feature.

Comment (ssoonmi) [reply frontend/src/scenes/surveys/surveyLogic.test.ts:63]: Actually, since there `choices` field is just a JSON string in the database, it can just be converted to an array of objects instead of an array of strings by the API backend when retrieving, creating, or updating a survey.

When creating or updating a survey:
```python
choices = raw_question.get("choices")
if choices and not isinstance(choices, list):
  raise serializers.ValidationError("Question choices must be a list of strings")
# if each choice is an instance of a string, convert it to an object before saving to database
```

When retrieving surveys:
```python
surveys = SurveyAPISerializer(
        Survey.objects.filter(team_id=team.id).exclude(archived=True).select_related("linked_flag", "targeting_flag"),
        many=True,
).data
# convert each survey's choices to an array of objects instead of an array of strings
```

If you set up the backend to migrate the legacy survey data like this, the frontend will not have to have logic to handle both cases where choices can be an array of objects or an array of strings. The frontend will always get an array of objects. So you can isolate the location where you may introduce potential bugs to the backend safely while being able to add more potential properties to a choice in the future (or even having multiple open-ended choices in a question).

Comment (neilkakkar) [reply frontend/src/scenes/surveys/surveyLogic.test.ts:63]: the problem is that posthog-js consumes this as well, and every API client that creates their own surveys, so changing structure will be a breaking change. It isn't limited to our backend-frontend communication, but agree that when we do go this route, the backend can do the backwards-compatible conversion

Comment (ssoonmi) [reply frontend/src/scenes/surveys/surveyLogic.test.ts:63]: Ah I see, so because the consumers of the API isn't just limited to the PostHog frontend, the API must keep the original data structure of the choices field to whatever was set by the original consumer. So the logic for converting choices to an array of objects would have to happen on the consumer side. Should it convert the choices to be an array of objects for any survey updated in the frontend even if it was originally as an array of strings? Or should only new surveys have save choices as an array of objects?

Comment (liyiy) [reply frontend/src/scenes/surveys/surveyLogic.test.ts:63]: @neilkakkar I think the breaking change probably makes sense to do now, along with this new feature so it's easier in one go. I don't think there'll be too many more, if any, changes to add to the MCQ choices array object after this open text "other" field. And it'll also be nice to make this change earlier on rather than later when even more people are using it.

Comment (liyiy) [reply frontend/src/scenes/surveys/surveyLogic.test.ts:63]: Easiest to just implement the `has_open_choice` field right now then and then we'll worry about choices array deprecation as it comes later on 😅

**Comment by ssoonmi** [comment]
I updated this PR and the other one (https://github.com/PostHog/posthog-js/pull/901) to use `has_open_choice` field on a question to determine if the question has an open-ended choice or not. Survey creation/update is forced to only have a single open-ended choice, and the open-ended choice will always be the last choice in the list of choices.

File: frontend/src/types.ts
Comment (liyiy) [inline frontend/src/types.ts:2180]: maybe camel case since it's only ever used in the frontend? Like `lowerBoundLabel` in rating survey question

```suggestion
    hasOpenChoice?: boolean
```

**Comment by liyiy** [comment]
I think we don't need the border for when the user has text input in the "Other" option, it looks good otherwise!

**Comment by liyiy** [comment]
I would recommend feature flagging this so we can get it merged ASAP without it breaking much. I can't figure out how to push to this branch lol otherwise I would help you feature flag it myself, but it should be a simple add.

In your `SurveyEdit.tsx` file, where you are supposed to add the open ended button choice, include this conditional right before everything

```
featureFlags[FEATURE_FLAGS.SURVEYS_OPEN_CHOICE]
```

And then in `constants.tsx` where the feature flags are defined, add this line to the bottom of that object

```
SURVEYS_OPEN_CHOICE: 'surveys-open-choice', // owner: @ssoonmi, #team-feature-success
```

And that should be it! To continue testing it locally, all you'll have to do is create a feature flag with `surveys-open-choice` as the key, enabled, and rolled out to 100% of everyone

**Comment by ssoonmi** [comment]
Just put the 'Add open-ended choice' button behind the feature flag!

**Comment by ssoonmi** [comment]
By border, do you mean the `open-ended` footer under the choice that is open-ended in the SurveyEdit form?

**Comment by liyiy** [comment]
Just this!

Screenshot 2023-11-22 at 10 31 55 AM

**Comment by ssoonmi** [comment]
> Just this! Removed completely looks the best

Done!

**Comment by ssoonmi** [comment]
I'm breaking one of the Cypress CI tests that I made, but I can't run the Cypress tests locally to test anymore for some reason. I'll try to fix it now though.

**Comment by ssoonmi** [comment]
@liyiy would you know if the Cypress tests have all feature flags enabled?

**Comment by liyiy** [comment]
> @liyiy would you know if the Cypress tests have all feature flags enabled?

Oh yes sorry, the flag will need to be added to `e2e.ts` under the cypress folder

**Comment by ssoonmi** [comment]
I tried to fix the code to pass as much of the CI tests as I could: https://github.com/PostHog/posthog/pull/18258/commits/440e31fc9b6f6237d941582bf574e4933c70efdb

**Comment by liyiy** [comment]
It looks like the backend tests are failing unrelated, can you try merging master again? It should fix it

**Comment by ssoonmi** [comment]
You mean merging master into this branch?

**Review by liyiy — APPROVED** [review]
Awesome stuff!!!

**Pull request merged**
